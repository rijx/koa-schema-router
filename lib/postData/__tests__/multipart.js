const Busboy = require("busboy");
const { EventEmitter } = require("events");

const installMultipartHandler = require("../multipart");

jest.mock("busboy");

describe("Multipart", () => {
  test("basic use", async () => {
    const dummyBusboy = new EventEmitter();

    Busboy.mockImplementation(() => dummyBusboy);

    const dummyContext = {
      req: {
        pipe: jest.fn()
      },
      request: {
        headers: {}
      }
    };

    installMultipartHandler(dummyContext);

    setImmediate(() => {
      dummyBusboy.emit(
        "file",
        "test",
        "[Stream]",
        "test.png",
        "base64",
        "type/png"
      );
      dummyBusboy.emit("field", "hello", "world");
      dummyBusboy.emit("finish");
    });

    const dummyProcessors = {
      test: jest.fn()
    };

    await dummyContext.request.body(dummyProcessors);

    expect(dummyProcessors.test.mock.calls.length).toBe(1);

    expect({
      dummyProcessor: dummyProcessors.test.mock.calls[0],
      body: dummyContext.request.body
    }).toMatchSnapshot();
  });

  test("missing processor", async () => {
    const dummyBusboy = new EventEmitter();

    Busboy.mockImplementation(() => dummyBusboy);

    const dummyContext = {
      req: {
        pipe: jest.fn()
      },
      request: {
        headers: {}
      }
    };

    installMultipartHandler(dummyContext);

    setImmediate(() => {
      dummyBusboy.emit(
        "file",
        "test",
        "[Stream]",
        "test.png",
        "base64",
        "type/png"
      );
      dummyBusboy.emit("field", "hello", "world");
      dummyBusboy.emit("finish");
    });

    const dummyProcessors = {};

    try {
      await dummyContext.request.body(dummyProcessors);

      throw new Error("Should have gone into catch block");
    } catch (err) {
      expect(err).toMatchSnapshot();
    }
  });
});
