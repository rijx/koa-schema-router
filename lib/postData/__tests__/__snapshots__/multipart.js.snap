// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Multipart basic use 1`] = `
Object {
  "body": Object {
    "hello": "world",
    "test": undefined,
  },
  "dummyProcessor": Array [
    Object {
      "contentType": "type/png",
      "encoding": "base64",
      "fileName": "test.png",
      "stream": "[Stream]",
    },
  ],
}
`;

exports[`Multipart missing processor 1`] = `[Error: No file processor for test]`;
